### PR TITLE
planner: firstrow agg func should infer return type as nullable for EQAll special case

### LINKED ISSUE #34584: Planner inferred wrong return type for firstrow agg function
Smaller repo:
```sql
create table foo(a int, b int);
insert into foo values(0, NULL);
alter table foo set tiflash replica 1;


create table bar(a double not null, b decimal(65,0)  not null);
insert into bar values(0, 0);
alter table bar set tiflash  replica 1;

select * from foo where a=all(select a from bar where bar.b=foo.b);
ERROR 1105 (HY000): other error for mpp stream: DB::Exception: Cannot convert NULL value to non-Nullable type

select * from foo where a=all(select 1 from bar where bar.b=foo.b);
ERROR 1105 (HY000): other error for mpp stream: DB::Exception: Cannot convert NULL value to non-Nullable type

explain select * from foo where a=all(select a from bar where bar.b=foo.b);
+------------------------------------------+---------+--------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                       | estRows | task         | access object | operator info                                                                                                                                                                                      |
+------------------------------------------+---------+--------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Projection_11                            | 1.00    | root         |               | test.foo.a, test.foo.b                                                                                                                                                                             |
| └─Apply_13                               | 1.00    | root         |               | CARTESIAN inner join, other cond:or(and(le(Column#8, 1), and(eq(cast(test.foo.a, double BINARY), Column#7), if(ne(Column#9, 0), NULL, 1))), or(eq(Column#10, 0), if(isnull(test.foo.a), NULL, 0))) |
|   ├─TableReader_15(Build)                | 1.00    | root         |               | data:TableFullScan_14                                                                                                                                                                              |
|   │ └─TableFullScan_14                   | 1.00    | cop[tikv]    | table:foo     | keep order:false, stats:pseudo                                                                                                                                                                     |
|   └─TableReader_34(Probe)                | 1.00    | root         |               | data:ExchangeSender_33                                                                                                                                                                             |
|     └─ExchangeSender_33                  | 1.00    | mpp[tiflash] |               | ExchangeType: PassThrough                                                                                                                                                                          |
|       └─Projection_29                    | 1.00    | mpp[tiflash] |               | Column#7, Column#8, Column#9, Column#10                                                                                                                                                            |
|         └─HashAgg_30                     | 1.00    | mpp[tiflash] |               | funcs:firstrow(Column#13)->Column#7, funcs:count(distinct test.bar.a)->Column#8, funcs:sum(Column#14)->Column#9, funcs:sum(Column#15)->Column#10                                                   |
|           └─ExchangeReceiver_32          | 1.00    | mpp[tiflash] |               |                                                                                                                                                                                                    |
|             └─ExchangeSender_31          | 1.00    | mpp[tiflash] |               | ExchangeType: PassThrough                                                                                                                                                                          |
|               └─HashAgg_19               | 1.00    | mpp[tiflash] |               | group by:test.bar.a, funcs:firstrow(test.bar.a)->Column#13, funcs:sum(0)->Column#14, funcs:count(1)->Column#15                                                                                     |
|                 └─Selection_28           | 0.80    | mpp[tiflash] |               | eq(test.bar.b, cast(test.foo.b, decimal(20,0) BINARY))                                                                                                                                             |
|                   └─TableFullScan_27     | 1.00    | mpp[tiflash] | table:bar     | keep order:false, stats:pseudo                                                                                                                                                                     |
+------------------------------------------+---------+--------------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

But in Tikv, it is getting expected result, same with MySQL:
```sql
mysql> set tidb_allow_mpp=0;
Query OK, 0 rows affected (0.19 sec)

mysql> select * from foo where a=all(select a from bar where bar.b=foo.b);
+------+------+
| a    | b    |
+------+------+
|    0 | NULL |
+------+------+
1 row in set (0.19 sec)

mysql> explain select * from foo where a=all(select a from bar where bar.b=foo.b);
+--------------------------------+---------+-----------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                             | estRows | task      | access object | operator info                                                                                                                                                                                      |
+--------------------------------+---------+-----------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Projection_11                  | 1.00    | root      |               | test.foo.a, test.foo.b                                                                                                                                                                             |
| └─Apply_13                     | 1.00    | root      |               | CARTESIAN inner join, other cond:or(and(le(Column#8, 1), and(eq(cast(test.foo.a, double BINARY), Column#7), if(ne(Column#9, 0), NULL, 1))), or(eq(Column#10, 0), if(isnull(test.foo.a), NULL, 0))) |
|   ├─TableReader_15(Build)      | 1.00    | root      |               | data:TableFullScan_14                                                                                                                                                                              |
|   │ └─TableFullScan_14         | 1.00    | cop[tikv] | table:foo     | keep order:false, stats:pseudo                                                                                                                                                                     |
|   └─StreamAgg_19(Probe)        | 1.00    | root      |               | funcs:firstrow(test.bar.a)->Column#7, funcs:count(distinct test.bar.a)->Column#8, funcs:sum(0)->Column#9, funcs:count(1)->Column#10                                                                |
|     └─TableReader_22           | 0.80    | root      |               | data:Selection_21                                                                                                                                                                                  |
|       └─Selection_21           | 0.80    | cop[tikv] |               | eq(test.bar.b, cast(test.foo.b, decimal(20,0) BINARY))                                                                                                                                             |
|         └─TableFullScan_20     | 1.00    | cop[tikv] | table:bar     | keep order:false, stats:pseudo                                                                                                                                                                     |
+--------------------------------+---------+-----------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by fixdb**
Issue Number: close #34584

Problem Summary:
Before this patch, the firstrow agg function infer the return type same with
the argument data type, which is wrong when the arg is not null during the
process planner decorrelate EQAll subqueries, and may cause error on TiFlash
engine.

Currently, firstrow agg function is treated like the exact representation of
aggregate group key, so the data type is the same with group key, even if the
group key is not null.

However, the return type of firstrow should be nullable, we clear the null flag
here instead of during invoking NewAggFuncDesc, in order to keep compatibility
with the existing presumption that the return type firstrow does not change
nullability, whatsoever.

This also can be testified by commit https://github.com/pingcap/tidb/commit/ad7102cdeedfcf5f014003f2c1579fc13d82e467, in which it stated that:
```
For all the aggregate functions except `first_row`, if we have an empty table
defined as t(a,b), `select agg(a) from t` would always return one row, while
`select agg(a) from t group by b` would return empty. For `first_row` which is
only used internally by tidb, `first_row(a)` would always return empty for
empty input now.
```

### Check List

Tests

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note
```release-note
None
```

### CODE REVIEW

**Comment by Reminiscent** [comment]
Some test cases are failed. Please fix them.

File: expression/aggregation/base_func.go [hunk 1]
```diff
@@ -286,7 +286,7 @@ func (a *baseFuncDesc) typeInfer4MaxMin(ctx sessionctx.Context) {
 		a.Args[0] = expression.BuildCastFunction(ctx, a.Args[0], tp)
 	}
 	a.RetTp = a.Args[0].GetType()
-	if (a.Name == ast.AggFuncMax || a.Name == ast.AggFuncMin) && a.RetTp.GetType() != mysql.TypeBit {
+	if (a.Name == ast.AggFuncMax || a.Name == ast.AggFuncMin || a.Name == ast.AggFuncFirstRow) && a.RetTp.GetType() != mysql.TypeBit {
```
Comment (Reminiscent) [inline expression/aggregation/base_func.go:289]: Just wondering why the `TypeBit` is excluded?

**Review by Reminiscent — COMMENTED** [review]
Please fix the failed test cases. Rest LGTM

**Review by qw4990 — COMMENTED** [review]
LGTM, please fix failed CI cases.

File: expression/aggregation/base_func.go (hunk 1 above)
Comment (fzhedu) [inline expression/aggregation/base_func.go:289]: does other agg functions have such problems?

Comment (fixdb) [reply expression/aggregation/base_func.go:289]: After a second thought, I think TypeBit should not be excluded.

Comment (fixdb) [reply expression/aggregation/base_func.go:289]: I think others are good.

File: cmd/explaintest/r/tpch.result [hunk 2]
```diff
@@ -988,7 +989,8 @@ Projection	1.00	root		div(Column#46, 7.0)->Column#47
     └─HashAgg(Probe)	9943040.00	root		group by:tpch.lineitem.l_partkey, funcs:avg(Column#50, Column#51)->Column#44, funcs:firstrow(tpch.lineitem.l_partkey)->tpch.lineitem.l_partkey
       └─TableReader	9943040.00	root		data:HashAgg
         └─HashAgg	9943040.00	cop[tikv]		group by:tpch.lineitem.l_partkey, funcs:count(tpch.lineitem.l_quantity)->Column#50, funcs:sum(tpch.lineitem.l_quantity)->Column#51
-          └─TableFullScan	300005811.00	cop[tikv]	table:lineitem	keep order:false
+          └─Selection	300005811.00	cop[tikv]		not(isnull(tpch.lineitem.l_partkey))
```
Comment (fixdb) [inline cmd/explaintest/r/tpch.result:992]: This should not be there. l_partkey is not null.

**Review by qw4990 — APPROVED** [review]

**Review by fzhedu — APPROVED** [review]
LGTM

**Pull request merged**